### ISSUE TITLE
Synchronization issues 1.61 and 1.62 versions

### ISSUE THREAD

**Issue by Gilis95** [bcgit/bc-java#548]
Hello, we're experiencing performance issues using latest BC releases. The problem is that most of the threads are stucked on random generation. To me it seems like there is some kind of a dead lock situation, I am attaching part of thread dump:
```
"requestExecutor247" - Thread t@944
   java.lang.Thread.State: RUNNABLE
        at java.lang.Thread.setPriority0(Native Method)
        at java.lang.Thread.setPriority(Thread.java:1095)
        at java.lang.Thread.init(Thread.java:417)
        at java.lang.Thread.init(Thread.java:349)
        at java.lang.Thread.<init>(Thread.java:464)
        at org.bouncycastle.jcajce.provider.drbg.DRBG$HybridSecureRandom$SignallingEntropySource.getEntropy(Unknown Source)
        at org.bouncycastle.crypto.prng.drbg.HMacSP800DRBG.getEntropy(Unknown Source)
        at org.bouncycastle.crypto.prng.drbg.HMacSP800DRBG.reseed(Unknown Source)
        at org.bouncycastle.crypto.prng.SP800SecureRandom.reseed(Unknown Source)
        - locked <fa8ec69> (a org.bouncycastle.crypto.prng.SP800SecureRandom)
        at org.bouncycastle.jcajce.provider.drbg.DRBG$HybridSecureRandom.generateSeed(Unknown Source)
        at org.bouncycastle.crypto.prng.BasicEntropySourceProvider$1.getEntropy(Unknown Source)
        at org.bouncycastle.crypto.prng.EntropyUtil.generateSeed(Unknown Source)
        at org.bouncycastle.crypto.prng.SP800SecureRandom.generateSeed(Unknown Source)
        at org.bouncycastle.jcajce.provider.drbg.DRBG$Default.engineGenerateSeed(Unknown Source)
        at java.security.SecureRandom.generateSeed(SecureRandom.java:533)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaTlsCryptoProvider$NonceEntropySource$NonceEntropySourceSpi.engineGenerateSeed(Unknown Source)
        at java.security.SecureRandom.generateSeed(SecureRandom.java:533)
        at org.bouncycastle.crypto.prng.BasicEntropySourceProvider$1.getEntropy(Unknown Source)
        at org.bouncycastle.crypto.prng.drbg.HashSP800DRBG.getEntropy(Unknown Source)
        at org.bouncycastle.crypto.prng.drbg.HashSP800DRBG.<init>(Unknown Source)
        at org.bouncycastle.crypto.prng.SP800SecureRandomBuilder$HashDRBGProvider.get(Unknown Source)
        at org.bouncycastle.crypto.prng.SP800SecureRandom.nextBytes(Unknown Source)
        - locked <56a1d17c> (a org.bouncycastle.crypto.prng.SP800SecureRandom)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaNonceGenerator.generateNonce(Unknown Source)
        at org.bouncycastle.tls.TlsProtocol.createRandomBlock(Unknown Source)
        at org.bouncycastle.tls.TlsClientProtocol.sendClientHelloMessage(Unknown Source)
        at org.bouncycastle.tls.TlsClientProtocol.beginHandshake(Unknown Source)
        at org.bouncycastle.tls.TlsClientProtocol.connect(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
        - locked <657dda72> (a org.bouncycastle.jsse.provider.ProvSSLSocketDirect)
        at com.sun.jndi.ldap.Connection.createSocket(Connection.java:393)
        at com.sun.jndi.ldap.Connection.<init>(Connection.java:215)
        at com.sun.jndi.ldap.LdapClient.<init>(LdapClient.java:137)
        at com.sun.jndi.ldap.LdapClient.getInstance(LdapClient.java:1609)
        at com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2749)
        at com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:319)
        at com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)



"requestExecutor208" - Thread t@854
   java.lang.Thread.State: BLOCKED
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaTlsCryptoProvider$NonceEntropySource$NonceEntropySourceSpi.engineNextBytes(Unknown Source)
        - waiting to lock <529ecb33> (a org.bouncycastle.jcajce.provider.digest.SHA512$Digest) owned by "requestExecutor204" t@869
        at java.security.SecureRandom.nextBytes(SecureRandom.java:468)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaNonceGenerator.<init>(Unknown Source)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaTlsCrypto.createNonceGenerator(Unknown Source)
        at org.bouncycastle.tls.AbstractTlsContext.createNonceGenerator(Unknown Source)
        at org.bouncycastle.tls.AbstractTlsContext.<init>(Unknown Source)
        at org.bouncycastle.tls.TlsClientContextImpl.<init>(Unknown Source)
        at org.bouncycastle.tls.TlsClientProtocol.connect(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
        - locked <6df2671b> (a org.bouncycastle.jsse.provider.ProvSSLSocketDirect)
        at com.sun.jndi.ldap.Connection.createSocket(Connection.java:393)
        at com.sun.jndi.ldap.Connection.<init>(Connection.java:215)
        at com.sun.jndi.ldap.LdapClient.<init>(LdapClient.java:137)
        at com.sun.jndi.ldap.LdapClient.getInstance(LdapClient.java:1609)
        at com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2749)
        at com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:319)
        at com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)
        at com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
        at com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:153)
        at com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:83)
        at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:684)
        at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
        at javax.naming.InitialContext.init(InitialContext.java:244)


"requestExecutor204" - Thread t@869
   java.lang.Thread.State: BLOCKED
        at org.bouncycastle.crypto.prng.SP800SecureRandom.nextBytes(Unknown Source)
        - waiting to lock <741266b1> (a org.bouncycastle.crypto.prng.SP800SecureRandom) owned by "requestExecutor73" t@437
        at org.bouncycastle.jcajce.provider.drbg.DRBG$Default.engineNextBytes(Unknown Source)
        at java.security.SecureRandom.nextBytes(SecureRandom.java:468)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaTlsCryptoProvider$NonceEntropySource$NonceEntropySourceSpi.engineNextBytes(Unknown Source)
        - locked <529ecb33> (a org.bouncycastle.jcajce.provider.digest.SHA512$Digest)
        at java.security.SecureRandom.nextBytes(SecureRandom.java:468)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaNonceGenerator.<init>(Unknown Source)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaTlsCrypto.createNonceGenerator(Unknown Source)
        at org.bouncycastle.tls.AbstractTlsContext.createNonceGenerator(Unknown Source)
        at org.bouncycastle.tls.AbstractTlsContext.<init>(Unknown Source)
        at org.bouncycastle.tls.TlsClientContextImpl.<init>(Unknown Source)
        at org.bouncycastle.tls.TlsClientProtocol.connect(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
        - locked <13773803> (a org.bouncycastle.jsse.provider.ProvSSLSocketDirect)
        at com.sun.jndi.ldap.Connection.createSocket(Connection.java:393)
        at com.sun.jndi.ldap.Connection.<init>(Connection.java:215)
        at com.sun.jndi.ldap.LdapClient.<init>(LdapClient.java:137)
        at com.sun.jndi.ldap.LdapClient.getInstance(LdapClient.java:1609)
        at com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2749)
        at com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:319)
        at com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:192)
        at com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:210)
        at com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:153)
        at com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:83)
        at javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:684)
        at javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:313)
        at javax.naming.InitialContext.init(InitialContext.java:244)
        at javax.naming.InitialContext.<init>(InitialContext.java:216)

"requestExecutor73" - Thread t@437
   java.lang.Thread.State: BLOCKED
        at org.bouncycastle.crypto.prng.SP800SecureRandom.nextBytes(Unknown Source)
        - waiting to lock <fa8ec69> (a org.bouncycastle.crypto.prng.SP800SecureRandom) owned by "requestExecutor247" t@944
        at org.bouncycastle.jcajce.provider.drbg.DRBG$HybridSecureRandom.generateSeed(Unknown Source)
        at org.bouncycastle.crypto.prng.BasicEntropySourceProvider$1.getEntropy(Unknown Source)
        at org.bouncycastle.crypto.prng.drbg.HashSP800DRBG.getEntropy(Unknown Source)
        at org.bouncycastle.crypto.prng.drbg.HashSP800DRBG.reseed(Unknown Source)
        at org.bouncycastle.crypto.prng.drbg.HashSP800DRBG.generate(Unknown Source)
        at org.bouncycastle.crypto.prng.SP800SecureRandom.nextBytes(Unknown Source)
        - locked <741266b1> (a org.bouncycastle.crypto.prng.SP800SecureRandom)
        at org.bouncycastle.jcajce.provider.drbg.DRBG$Default.engineNextBytes(Unknown Source)
        at java.security.SecureRandom.nextBytes(SecureRandom.java:468)
        at java.util.UUID.randomUUID(UUID.java:145)

```

41threads are blocked on retrieving lock on object `529ecb33`, mutual stacktrace:
```
"requestExecutor444" - Thread t@1253
   java.lang.Thread.State: BLOCKED
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaTlsCryptoProvider$NonceEntropySource$NonceEntropySourceSpi.engineNextBytes(Unknown Source)
        - waiting to lock <529ecb33> (a org.bouncycastle.jcajce.provider.digest.SHA512$Digest) owned by "requestExecutor204" t@869
        at java.security.SecureRandom.nextBytes(SecureRandom.java:468)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaNonceGenerator.<init>(Unknown Source)
        at org.bouncycastle.tls.crypto.impl.jcajce.JcaTlsCrypto.createNonceGenerator(Unknown Source)
        at org.bouncycastle.tls.AbstractTlsContext.createNonceGenerator(Unknown Source)
        at org.bouncycastle.tls.AbstractTlsContext.<init>(Unknown Source)
        at org.bouncycastle.tls.TlsClientContextImpl.<init>(Unknown Source)
        at org.bouncycastle.tls.TlsClientProtocol.connect(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
        at org.bouncycastle.jsse.provider.ProvSSLSocketDirect.startHandshake(Unknown Source)
```

**Comment by kimmerin** [comment]
What SecureRandom is provided during the instantiation of JcaTlsCryptoProvider? If that happened to be

    SecureRandom.getInstanceStrong()

and your test runs on a Linux system without GUI you run into the trap that e.g. is described at [https://www.2uo.de/myths-about-urandom](url). That instance of SecureRandom uses /dev/random that blocks if it decides not having enough entropy.

Use a different SecureRandom instance or make sure the system gets enough entropy to avoid blocking events like these e.g by installing packages that get around this problem (e.g. rng-tools) or have some hardware based source of randomness.

**Comment by Gilis95** [comment]
We haven't provide `org.bouncycastle.drbg.entropysource` variable, which means `org.bouncycastle.jcajce.provider.drbg.DRBG#HybridSecureRandom` is being used as SecureRandom provider. As it seems this class is using base secure random instance, that is instantiated using `getInstanceStrong` method. This is happening inside `DRBG#createInitialEntropySource`. I am going to test exporting the variable and I'll close if issue doesn't persist anymore, thanks.

**Comment by Gilis95** [comment]
I have provided the variable and we're no longer seeing threads blocked on places listed above.
Now we're hitting next blocking place:
```
"requestExecutor447" - Thread t@1268
   java.lang.Thread.State: BLOCKED
        at org.bouncycastle.crypto.prng.SP800SecureRandom.nextBytes(Unknown Source)
        - waiting to lock <74bad4e4> (a org.bouncycastle.crypto.prng.SP800SecureRandom) owned by "requestExecutor474" t@1271
        at org.bouncycastle.jcajce.provider.drbg.DRBG$Default.engineNextBytes(Unknown Source)
        at java.security.SecureRandom.nextBytes(SecureRandom.java:468)
        at java.util.UUID.randomUUID(UUID.java:145)
``` 
There are synchronized blocks inside `SP800SecureRandom`, where our application blocks now.

**Comment by Gilis95** [comment]
I forgot to mention, with EntropyProvider, we're providing our EntropySource, that is referring to `NativePRNGNonBlocking`. In other words `/dev/urandom` is being used for generation.

**Comment by kimmerin** [comment]
What's requestExecutor474 doing (the one holding the lock for <74bad4e4>)?

**Comment by Gilis95** [comment]
It's been waiting on random generation. However the problem seems to be not the lack of provided variable pointing to `EntropySourceProvider`, but the path to random generation script. It's located into `java.security` file, we've filled in next properties `securerandom.source=file:/dev/urandom` and `securerandom.strongAlgorithms=NativePRNGNonBlocking:SUN`. It looks pretty normal and as it should work. When `getInstanceStrong` is called `algorithm NativePRNGNonBlocking` and provider `SUN` will be used, random source SHOULD be `/dev/urandom`. However there is surprise on java side, that won't be fixed. Here it is the big -> https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6202721. Even though you've provided `/dev/urandom`, `/dev/random` will be used. The only way to workaround this, is to provide `/dev/./urandom`. Closing this, as not an issue.
